### PR TITLE
replace label images with segmentation masks

### LINKED ISSUE #875: Propose a format for segmentation results
# Cell Segmentation and format specification

We require a 3-d instance segmentation format, which should be possible to use to annotate IntensityTables. There are an adequate number of already-existing formats from the instance segmentation literature that can be grouped into two categories:

1. Instance masks, either binary or probabilistic, which can be compressed
2. polygons or meshes

Both types have robust library implementations in python (see examples below). Compressed binary masks and polygons appear possible to interconvert (see e.g. https://github.com/facebookresearch/Detectron/issues/100). It seems like either approach might be adequate for our purposes, as both are nicely supported by existing libraries. 

**Use cases**:
1. Geometric intersection of segmented objects with identified spots. 
2. Geometric intersection of segmented objects with other objects

## Examples of instance masks: 
https://github.com/matterport/Mask_RCNN/issues/297
http://scikit-image.org/docs/dev/api/skimage.measure.html?highlight=profile#skimage.measure.label

## Examples of polygons in 2d
https://pypi.org/project/Shapely/
https://pythonhosted.org/planar/overview.html
https://github.com/freeman-lab/regional
http://scikit-image.org/docs/dev/api/skimage.measure.html
http://geopandas.org/

## Examples of polygons in 3d
https://pymesh.readthedocs.io/en/latest/

## Formats for serialization
The formats for serialization appear to be driven by either annotated datasets or software for annotation. 
http://cocodataset.org/#format-results and https://github.com/cocodataset/cocoapi
http://www.robots.ox.ac.uk/~vgg/software/via/, which saves data in a json or csv format that looks like the following: 
```
#filename,file_size,file_attributes,region_count,region_id,region_shape_attributes,region_attributes
swan_in_geneve.jpg,62201,"{""caption"":""Swan in a lake""}",1,0,"{""name"":""polygon"",""all_points_x"":[119,102,196,406,395,433,413,336,332,247,164,119],""all_points_y"":[175,218,285,288,232,227,140,138,210,182,195,175]}","{""name"":""Swan"",""color"":""white""}"
```

## User requirements: 
@ambrosejcarr is actively doing research to determine whether we have additional requirements. 

**Watershed**: 
- Each pixel is hard-assigned to at most one cell and is stored in memory as a label image where each cell is assigned a numerical value, and each pixels is assigned the value corresponding to the cell, or 0 if it is not part of a cell. 

**Voronoi Tessellation**:
- Each pixel is hard-assigned to exactly one cell. 

**Kenneth Harris' method**: 
- Each pixel is hard-assigned to at most one cell, in initial segmentation, and is stored in memory as a label image where each cell is assigned a numerical value, and each pixels is assigned the value corresponding to the cell, or 0 if it is not part of a cell. 
- Spots are assigned probabilistically, using a sparse array containing probabilities for the top few cells

## Suggested Direction: 
Binary segmentation masks support most use cases and can be easily converted into polygons as necessary, although the converse is not true. Based on discussions with experts, it seems like masks are a better choice for computationally generated masks, and are closer to the image, and therefore easier to apply and visualize. As a result, I suggest we start with an implementation that focuses on using masks. 

## Implementation requirements: 
1. SpaceTx specification for a binary segmentation mask. Need not enable probabilistic or overlapping representations, but the implementation should be aware of a potential future need (suggest 1 sparse mask per object, not 1 mask per image that contains all objects). 
2. Starfish object to represent a set of segmented cells that meets the requirements of the specification. This object should be usable either with a single field of view, or on a stitched (very large) composite image. 
3. On-disk storage format for the segmentation mask
4. Update starfish watershed implementation to emit the format (see also #998 for a suggested improvement)

### PR DESCRIPTION
**Pull request by kne42**
closes #875

### CODE REVIEW

File: starfish/segmentation_mask.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+        Returns
+        -------
+        masks : SegmentationMaskCollection
+            Masks generated from the label image.
+        """
+        props = regionprops(label_image)
+
+        if label_image.ndim == 2:
+            dims = AXES[1:]
```
Comment (ttung) [inline starfish/segmentation_mask.py]: see earlier comments.

File: starfish/segmentation_mask.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        elif label_image.ndim == 3:
+            dims = AXES
+        else:
+            raise TypeError('expected 2- or 3-D image')
+
+        masks: List[xr.DataArray] = []
+
+        coords: Dict[str, Union[list, Tuple[str, list]]]
+
+        for label, prop in enumerate(props):
```
Comment (ttung) [inline starfish/segmentation_mask.py]: this code is very dense, and is probably worth some comments.

File: starfish/segmentation_mask.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+        path : str
+            Path of the directory to instantiate from.
+
+        Returns
+        -------
+        masks : SegmentationMaskCollection
+            Collection of segmentation masks.
+        """
+        masks = []
+        for p in os.listdir(path):
```
Comment (ttung) [inline starfish/segmentation_mask.py]: from a file format issue, I don't think this is ideal.  I would much prefer that we have a strategy to group these all together, whether it be a tar file or a json file to stitch them all together.

File: starfish/segmentation_mask.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+        return cls(masks)
+
+    def save(self, path: str, overwrite: bool = False):
+        """Save the segmentation masks to disk.
+
+        Parameters
+        ----------
+        path : str
+            Path of the directory to write to.
+        overwrite : bool, optional
```
Comment (ttung) [inline starfish/segmentation_mask.py]: I think the pattern used thus far in this repo is to do:

```
overwrite : bool
    Whether to overwrite the directory if it exists. (default: False)
```

If you don't feel strongly about it, let's stick with this convention.  If you do feel strongly about it, let's discuss. :)

File: starfish/segmentation_mask.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+            Whether to overwrite the directory if it exists.
+        """
+        try:
+            os.mkdir(path)
```
Comment (ttung) [inline starfish/segmentation_mask.py]: consider using os.makedirs, which more or less deals with this problem. :)

File: starfish/test/test_segmentation_mask.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        validate_segmentation_mask(bad)
+
+    bad = xr.DataArray([True],
+                       dims=('x'),
+                       coords=dict(x=[0],
+                                   xc=('x', [0.5])))
+    with pytest.raises(TypeError):
+        validate_segmentation_mask(bad)
+
+    bad = xr.DataArray([[True]],
```
Comment (shanaxel42) [inline starfish/test/test_segmentation_mask.py]: good names :)

File: starfish/segmentation_mask.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+        if coord not in arr.coords:
+            raise TypeError(f"no coordinate '{coord}'")
+
+
+class SegmentationMaskCollection:
+    """Collection of binary segmentation masks with a list-like access pattern.
+
+    Parameters
+    ----------
+    masks : list of xr.DataArray
```
[the remark this answers is not in the document]
Comment (kne42) [reply starfish/segmentation_mask.py]: The whole collection will effectively serve as a proxy to these. Do you think it's better practice to expose this as a public attribute?

What I would normally/prefer to do is not create a class at all and just use a functional approach which is much cleaner IMO (outside of type hints) but I wanted this to blend in with the rest of the package which feels a bit class-happy tbh.

File: starfish/segmentation_mask.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+        Segmentation masks.
+    """
+    _masks: List[xr.DataArray]
```
[the remark this answers is not in the document]
Comment (kne42) [reply starfish/segmentation_mask.py]: No, just thought maybe people would want this to be typed idk

File: starfish/spots/_target_assignment/label.py [hunk 9]
```diff
@@ -19,26 +18,29 @@ def _add_arguments(cls, parser) -> None:
...
     @staticmethod
     def _assign(
-        label_image: np.ndarray,
+        masks: SegmentationMaskCollection,
         intensities: IntensityTable,
         in_place: bool,
     ) -> IntensityTable:
+        cell_ids = []
+
+        for spot in intensities:
```
[the remark this answers is not in the document]
Comment (kne42) [reply starfish/spots/_target_assignment/label.py]: For our test cases there aren't many, sure. If you think that there are users who (will) have a lot of spots/masks then we could look into improving the algorithm.

File: starfish/image/_segmentation/watershed.py [hunk 10]
```diff
@@ -49,7 +50,12 @@ def __init__(
         self.min_distance = min_distance
         self._segmentation_instance: Optional[_WatershedSegmenter] = None
 
-    def run(self, primary_images: ImageStack, nuclei: ImageStack, *args) -> np.ndarray:
+    def run(
+            self,
+            primary_images: ImageStack,
+            nuclei: ImageStack,
+            *args
```
[the remark this answers is not in the document]
Comment (kne42) [reply starfish/image/_segmentation/watershed.py]: This was added in #1107

**Comment by kne42** [comment]
Ok, thanks for all the comments! I've addressed them (I hope) so this should be ready for re-review :)

**Comment by kne42** [comment]
There are also some nice features I could add such as enforcing unique label names which would open up the possibility to index by label or lazy load from disk.

**Comment by ttung** [comment]
> Ok, thanks for all the comments! I've addressed them (I hope) so this should be ready for re-review :)

when you want to ask for a re-review, you can do so using the github UI.  This way, it shows up on the reviewers' dashboards.

Screen Shot 2019-04-12 at 10 49 55 AM

File: starfish/segmentation_mask.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ttung) [reply starfish/segmentation_mask.py]: I think the constructor assigning this variable implicitly types the field.

File: starfish/segmentation_mask.py [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+            # subset from the full span of labels
+            for d, c in physical_ticks.items():
+                axis = d.value[0]
+                i = dims.index(axis)
+                coords[d.value] = (axis, c[prop.bbox[i]:prop.bbox[i + len(dims)]])
+
+            mask = xr.DataArray(prop.image,
+                                dims=dims,
+                                coords=coords,
+                                name=str(label + 1))
```
Comment (ttung) [inline starfish/segmentation_mask.py]: you could consider zero-padding the labels such that they sort nicely.

File: starfish/segmentation_mask.py [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+        return cls(masks)
+
+    @classmethod
+    def from_disk(cls, path: str) -> "SegmentationMaskCollection":
+        """Load the collection from disk.
+
+        Parameters
+        ----------
+        path : str
+            Path of the directory to instantiate from.
```
Comment (ttung) [inline starfish/segmentation_mask.py]: this needs to be updated.

File: starfish/segmentation_mask.py [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+        -------
+        masks : SegmentationMaskCollection
+            Collection of segmentation masks.
+        """
+        masks = []
+
+        with tarfile.open(path) as t:
+            for info in t.getmembers():
+                f = t.extractfile(info.name)
+                print(f)
```
Comment (ttung) [inline starfish/segmentation_mask.py]: remove?

**Pull request merged**